EIPS/eip-3540.md

```diff
@@ -47,38 +47,35 @@ In order to guarantee that every EOF-formatted contract in the state is valid, w
+If code starts with the `MAGIC`, it is considered to be EOF formatted, otherwise it is considered to be *legacy* code. For clarity, the `MAGIC` together with a version number *n* is denoted as the *EOFn prefix*, e.g. *EOF1 prefix*.
 
-The opcode `0xEF` is currently an undefined instruction, therefore: *It pops no stack items and pushes no stack items, and it causes an exceptional abort when executed.* This means *initcode* or already deployed *code* starting with this instruction will continue to abort execution.
+EOF-formatted contracts are created using new instructions which are introduced in a separate EIP.
+
+The opcode `0xEF` is currently an undefined instruction, therefore: *It pops no stack items and pushes no stack items, and it causes an exceptional abort when executed.* This means legacy *initcode* or already deployed legacy *code* starting with this instruction will continue to abort execution.
 
 Unless otherwised specified, all integers are encoded in big-endian byte order.
 
 ### Code validation
 
 We introduce *code validation* for new contract creation. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.
 
-At `block.number == HF_BLOCK` new contract creation is modified:
-
-- if *initcode* or *code* starts with the `MAGIC`, it is considered to be EOF formatted and will undergo validation specified in the following sections,
-- else if *code* starts with `0xEF`, creation continues to result in an exceptional abort (the rule introduced in EIP-3541),
-- otherwise code is considered *legacy code* and the following rules do not apply to it.
+Legacy code is not affected by EOF code validation.
 
-For a create transaction, if *initcode* or *code* is invalid, the contract creation results in an exceptional abort. Such a transaction is valid and may be included in a block. Therefore, the transaction sender's nonce is increased.
-
-For the `CREATE` and `CREATE2` instructions, if *initcode* or *code* is invalid, instructions' execution ends with the result `0` pushed on stack. The *initcode* validation happens just before its execution and validation failure is observable as if execution results in an exceptional abort. I.e. in case *initcode* or returned *code* is invalid the caller's nonce remains increased and all creation gas is deducted.
+Code validation is performed during contract creation, and is elaborated on in separate EIPs.
+The EOF format itself and its formal validation are described in the following sections.
 
 ### Container specification
 
 EOF container is a binary format with the capability of providing the EOF version number and a list of EOF sections.
 
-The container starts with the EOF header:
+The container starts with the EOF prefix:
 
 | description | length   | value      |                    |
 |-------------|----------|------------|--------------------|
 | magic       | 2-bytes  | 0xEF00     |                    |
```

Comment:
Can we add a ref to The MAGIC section below?

Reply:
As mentioned elsewhere, I think this is a tiny bit non-standard to link to the Rationale section bits. If done scrupulously for all the pieces of the spec, would overwhelm the spec itself.